fairscale/experimental/nn/mevo.py

```diff
@@ -0,0 +1,480 @@
+         ...
+
+       def forward():
+         w = self.trigger()
+         ... continue to use w ...
+       ```
+
+       As a resule, the trigger's backward hook will be called at the end of
+       the backward for the module that uses this trigger.
+    """
+
+    def __init__(self, linked_param: torch.Tensor):
+        super().__init__()
+        assert isinstance(linked_param, nn.Parameter)
+        self.trigger = nn.Parameter(torch.rand(1, dtype=linked_param.dtype))
+        self.trigger._linked_param = linked_param
+
+    def forward(self, *input: Any, **kwargs: Any) -> Any:
+        return BackwardTriggerFn.apply(self.trigger._linked_param, self.trigger)
+
+
+class MemoryEfficientVocabOutput(nn.Module):  # AKA. MEVO
+    """ Fused fc + softmax + nll_loss in a tiled fashion.
+
+        This uses much less memory but is quite a bit slower.
+
+    Args:
+        proj_weight (nn.Parameter):
+            Sharing this weight with an embedding layer.
+        tile_factor (int):
+            Number of splits to use on the input sequence and vocab dimensions.
+        reduction (str):
+            Reduction OP (sum or mean).
+    """
+
+    def __init__(self, proj_weight: nn.Parameter, tile_factor: int = 16, reduction: str = "sum"):
+        super().__init__()
+        self.proj_weight = proj_weight
+        self.tf_in, self.tf_w = tile_factor, tile_factor
+        self.fp_max = True
+        self.fp_sum = True  # This is esp. important when tensors are large. Otherwise, you get inf.
+        self.fp_target = True
+        self.log_softmax = True
+        self.reduction = reduction
+        assert self.reduction in ["sum", "mean"]
+        self.trigger = BackwardTrigger(self.proj_weight)
+        if DEBUG and dist.is_initialized() and dist.get_rank() == 0:
+            print(
+                f"DEBUG cfg tf_in={self.tf_in} tf_w={self.tf_w} fp_max={self.fp_max} "
+                f"fp_sum={self.fp_sum} fp_target={self.fp_target} log_softmax={self.log_softmax} "
+                f"reduction={self.reduction}"
+            )
+
+    def get_target_nlprob(
+        self, i: torch.Tensor, w: torch.Tensor, target: torch.Tensor, debase_max: torch.Tensor, exp_sums: torch.Tensor
+    ) -> torch.Tensor:
+        """Get target's negative log probability."""
+        target_score = TargetScoreFunction.apply(i, w, target, self)
+        prob = (target_score - debase_max).exp() / exp_sums
+        if self.log_softmax:
+            # lprob
+            prob = prob.log()
+        # nlprob, then sum over all tokens.
+        return -prob.sum()
+
+    def forward(self, *input: Any, **kwargs: Any) -> Any:
+        if DEBUG and dist.is_initialized() and dist.get_rank() == 0:
+            cur_mem = round(torch.cuda.memory_allocated() / 1024 / 1024)
+            mem = round(torch.cuda.max_memory_allocated() / 1024 / 1024)
+            print("DEBUG cur, peak", cur_mem, mem)
+        assert kwargs == {}
+        input, target = input
```

Comment:
I don't get that, why not declare them directly as inputs and get some help from mypy ?

Reply:
thanks for the review, Ben!

if I recall this correctly, the prototype for `forward` in mypy is actually the Any, Any one. That's why I used that to make mypy happier. If you know a way that I can use separate args more directly here without too much mypy magic, that'd be great!

Reply:
oh I typically ignore the mypy warning on this line ( #type: ignore), because I think that declaring inputs helps with readability (a lot of people will look at the interface, a lot more than the number of people who will look at the code in any case). The issue with the inputs goes with Pytorch not being typed in general I think, so this #type: ignore is a little ugly but I can live with it